File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/py_to_cpp/code_output/DiscountStrategy.cpp
#include <string>
#include <vector>
#include <functional>
#include <unordered_set>

struct Customer {
    std::string name;
    double fidelity;
};

struct Item {
    std::string product;
    int quantity;
    double price;
};

class DiscountStrategy {
public:
    Customer customer;
    std::vector<Item> cart;
    std::function<double(DiscountStrategy&)> promotion;

    DiscountStrategy(Customer customer, std::vector<Item> cart, std::function<double(DiscountStrategy&)> promotion = nullptr)
        : customer(std::move(customer)), cart(std::move(cart)), promotion(std::move(promotion)), _total(0.0) {
        _total = total();
    }

    double total() {
        _total = 0.0;
        for (const auto& item : cart) {
            _total += item.quantity * item.price;
        }
        return _total;
    }

    double due() {
        double discount = 0.0;
        if (promotion) {
            discount = promotion(*this);
        }
        return _total - discount;
    }

    static double FidelityPromo(DiscountStrategy& order) {
        return order.customer.fidelity >= 1000 ? order.total() * 0.05 : 0.0;
    }

    static double BulkItemPromo(DiscountStrategy& order) {
        double discount = 0.0;
        for (const auto& item : order.cart) {
            if (item.quantity >= 20) {
                discount += item.quantity * item.price * 0.1;
            }
        }
        return discount;
    }

    static double LargeOrderPromo(DiscountStrategy& order) {
        std::unordered_set<std::string> products;
        for (const auto& item : order.cart) {
            products.insert(item.product);
        }
        return products.size() >= 10 ? order.total() * 0.07 : 0.0;
    }

private:
    double _total;
};

File: data/derived_cot/rq1_traces/translation/glm5.1/cpp_to_py/code_output/DiscountStrategy.py
from enum import Enum, auto
from typing import Dict, List


class PromoType(Enum):
    FidelityPromo = auto()
    BulkItemPromo = auto()
    LargeOrderPromo = auto()
    NoPromo = auto()


class DiscountStrategy:
    def __init__(
        self,
        customer: Dict[str, float],
        cart: List[Dict[str, float]],
        promo: PromoType = PromoType.NoPromo,
    ):
        self.customer_ = customer
        self.cart_ = cart
        self.promo_ = promo

    def total(self) -> float:
        return sum(item["quantity"] * item["price"] for item in self.cart_)

    def due(self) -> float:
        return self.total() - self.promotion(self)

    def promotion(self, order: "DiscountStrategy") -> float:
        if self.promo_ == PromoType.FidelityPromo:
            return self.fidelity_discount()
        elif self.promo_ == PromoType.BulkItemPromo:
            return self.bulk_item_discount()
        elif self.promo_ == PromoType.LargeOrderPromo:
            return self.large_order_discount()
        else:
            return 0.0

    def fidelity_discount(self) -> float:
        total = 0.0
        fidelity = self.customer_["fidelity"]
        if fidelity >= 1000.0:
            total = 0.05 * self.total()
        return total

    def bulk_item_discount(self) -> float:
        discount = 0.0
        for item in self.cart_:
            if item["quantity"] >= 20:
                discount += item["quantity"] * item["price"] * 0.1
        return discount

    def large_order_discount(self) -> float:
        num_items = len(self.cart_)
        discount = 0.0
        if num_items >= 10:
            discount = 0.07 * self.total()
        return discount